components/speedreader/rust/lib/src/readability/src/extractor.rs

```diff
@@ -49,18 +76,178 @@ where
+                    | "dc:description"
+                    | "dcterm:description"
+                    | "og:description"
+                    | "weibo:article:description"
+                    | "weibo:webpage:description"
+                    | "twitter:description" => {
+                        meta_tags.description = Some(
+                            content
+                                .find(". ")
+                                .map(|pos| &content[..pos])
+                                .unwrap_or(content)
+                                .to_string(),
+                        );
+                    }
+                    "dc:creator" | "dcterm:creator" | "author" => {
+                        meta_tags.author = Some(content.to_string());
+                    }
+                    _ => (),
+                }
+            }
+        } else if let Some(charset) = data.attributes.borrow().get(local_name!("charset")) {
+            meta_tags.charset = Some(charset.to_string());
+        } else if data
+            .attributes
+            .borrow()
+            .get(local_name!("http-equiv"))
+            .map(|e| e.to_ascii_lowercase() == "content-type")
+            .unwrap_or(false)
+        {
+            if let Some(content) = data.attributes.borrow().get(local_name!("content")) {
+                if let Some(charset) = content.split("charset=").nth(1) {
+                    meta_tags.charset = Some(charset.trim().to_string());
+                }
+            }
+        }
+    }
+
+    let mut meta = meta_jsonld.merge(meta_tags);
+
+    if meta.title.is_empty() {
+        if let Some(handle) = meta_nodes
+            .into_iter()
+            .find(|d| d.as_element().unwrap().name.local == local_name!("title"))
+        {
+            dom::extract_text(&handle, &mut meta.title, true);
+        }
+    }
+
+    // Clean title of html encoded attributes
+    if !meta.title.is_empty() {
+        if let Ok(s) = decode_html(&meta.title) {
+            meta.title = DECODED_HTML_TAGS.replace_all(&s, "").into();
+        }
+    }
+
+    // Clean description of html encoded attributes
+    if let Some(ref description) = meta.description {
+        if let Ok(s) = decode_html(description) {
+            meta.description = Some(DECODED_HTML_TAGS.replace_all(&s, "").into());
+        }
+    }
+
+    meta
+}
+
 pub fn extract_dom<S: ::std::hash::BuildHasher>(
     mut dom: &mut Sink,
     url: &Url,
     features: &HashMap<String, u32, S>,
 ) -> Result<Product, std::io::Error> {
-    let mut meta = Meta::default();
     let handle = dom.document_node.clone();
 
+    let mut meta = extract_metadata(dom);
+    meta.title = clean_title(dom, meta.title);
+
     // extracts title (if it exists) pre-processes the DOM by removing script
     // tags, css, links
-    scorer::preprocess(&mut dom, handle.clone(), &mut meta);
-    meta.title = clean_title(dom, meta.title);
+    scorer::preprocess(&mut dom, handle.clone());
```

Comment:
is this second clone needed?  it seems like its a clone of a clone no?  Does it result in 2-3 copies of the same document in memory?

Reply:
these are of types `Rc` and the cl one function [increments a counter](https://doc.rust-lang.org/src/alloc/rc.rs.html#1366-1386)

Reply:
okay, i'll change this one that showed up in the diff, but i'll get the rest of them when i do `cargo clippy` fixes

Reply:
actually, i agree this is an issue, but can we table this one for now? mostly because `kuchiki` [exposes the type Rc<Node> as as a NodeRef](https://docs.rs/kuchiki/0.4.1/src/kuchiki/tree.rs.html#94). Using the standard `Rc::clone(..)` would mean some pretty ugly casting everywhere we do it, and the kuchiki codebase calls `.clone()`. I would need to think about it more.

Reply:
Sure, didn't mean for it to hold up this PR :sweat_smile:

I understand it's already that way all over the crate, so would be worth a separate look. I also think `NodeRef::clone(..)` should probably be clear enough rather than casting everywhere.